File: ts/core/src/parser/syntax/FuncScriptParser.GetKeyWordLiteral.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult } from "../FuncScriptParser.Main";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";

export function GetKeyWordLiteral(context: ParseContext, index: number): ParseResult {
    let parseNode: ParseNode | null = null;
    let i = GetLiteralMatch(context, index, "null").NextIndex;
    let literal: any;

    if (i > index) {
        literal = null;
    } else if ((i = GetLiteralMatch(context, index, "true").NextIndex) > index) {
        literal = true;
    } else if ((i = GetLiteralMatch(context, index, "false").NextIndex) > index) {
        literal = false;
    } else {
        return { ParseNode: null, NextIndex: index };
    }

    parseNode = new ParseNode(ParseNodeType.KeyWord, index, i - index);
    return { ParseNode: parseNode, NextIndex: i };
}

File: funcscript/parser/syntax/FuncScriptParser.GetKeyWordLiteral.cs
using FuncScript.Model;
namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        record GetKeyWordLiteralResult(object Literal, ParseNode ParseNode, int NextIndex)
            : ParseResult(ParseNode, NextIndex);

        static GetKeyWordLiteralResult GetKeyWordLiteral(ParseContext context, int index)
        {
            ParseNode parseNode = null;
            int i = GetLiteralMatch(context, index, "null").NextIndex;
            object literal;

            if (i > index)
            {
                literal = null;
            }
            else if ((i = GetLiteralMatch(context, index, "true").NextIndex) > index)
            {
                literal = true;
            }
            else if ((i = GetLiteralMatch(context, index, "false").NextIndex) > index)
            {
                literal = false;
            }
            else
            {
                return new GetKeyWordLiteralResult(null, null, index);
            }

            parseNode = new ParseNode(ParseNodeType.KeyWord, index, i - index);
            return new GetKeyWordLiteralResult(literal, parseNode, i);
        }
    }
}
